From: Folia Compatibility Patch
Subject: [PATCH 0005] FancyMessage: Replace dispatchCommand logic with ComponentSerializer.parse()

AE PATCH REASON: FancyMessage uses legacy JSON conversion incompatible with modern Adventure API
AE PATCH FIX: Replace dispatchCommand logic with ComponentSerializer.parse()
AE PATCH RISK: LOW
AE PATCH EXCLUDES: net/advancedplugins/ae/utils/fanciful/FancyMessage.class, net/advancedplugins/ae/utils/fanciful/FancyMessage$*.class

---
diff --git a/net/advancedplugins/ae/utils/fanciful/FancyMessage.java b/net/advancedplugins/ae/utils/fanciful/FancyMessage.java
index d7f2fff..c077763 100644
--- a/net/advancedplugins/ae/utils/fanciful/FancyMessage.java
+++ b/net/advancedplugins/ae/utils/fanciful/FancyMessage.java
@@ -36,12 +36,12 @@ public class FancyMessage implements JsonRepresentedObject, Cloneable, Iterable<
    private String jsonString;
    private boolean dirty;
 
-   public FancyMessage(String var1) {
-      this(TextualComponent.rawText(ColorUtils.format(var1)));
+   public FancyMessage(String firstPartText) {
+      this(TextualComponent.rawText(ColorUtils.format(firstPartText)));
    }
 
-   private FancyMessage(TextualComponent var1) {
-      this.messageParts.add(new MessagePart(var1));
+   private FancyMessage(TextualComponent firstPartText) {
+      this.messageParts.add(new MessagePart(firstPartText));
       this.jsonString = null;
       this.dirty = false;
    }
@@ -50,262 +50,262 @@ public class FancyMessage implements JsonRepresentedObject, Cloneable, Iterable<
       this((TextualComponent)null);
    }
 
-   public static FancyMessage deserialize(Map<String, Object> var0) {
-      FancyMessage var1 = new FancyMessage();
-      var1.messageParts = (List<MessagePart>)var0.get("messageParts");
-      var1.jsonString = var0.containsKey("JSON") ? var0.get("JSON").toString() : null;
-      var1.dirty = !var0.containsKey("JSON");
-      return var1;
+   public static FancyMessage deserialize(Map<String, Object> serialized) {
+      FancyMessage msg = new FancyMessage();
+      msg.messageParts = (List<MessagePart>)serialized.get("messageParts");
+      msg.jsonString = serialized.containsKey("JSON") ? serialized.get("JSON").toString() : null;
+      msg.dirty = !serialized.containsKey("JSON");
+      return msg;
    }
 
-   private static FancyMessage deserialize(String var0) {
-      JsonObject var1 = _stringParser.parse(var0).getAsJsonObject();
-      JsonArray var2 = var1.getAsJsonArray("extra");
-      FancyMessage var3 = new FancyMessage();
-      var3.messageParts.clear();
+   private static FancyMessage deserialize(String json) {
+      JsonObject serialized = _stringParser.parse(json).getAsJsonObject();
+      JsonArray extra = serialized.getAsJsonArray("extra");
+      FancyMessage returnVal = new FancyMessage();
+      returnVal.messageParts.clear();
 
-      for (JsonElement var5 : var2) {
-         MessagePart var6 = new MessagePart();
-         JsonObject var7 = var5.getAsJsonObject();
+      for (JsonElement mPrt : extra) {
+         MessagePart component = new MessagePart();
+         JsonObject messagePart = mPrt.getAsJsonObject();
 
-         for (Entry var9 : var7.entrySet()) {
-            if (TextualComponent.isTextKey((String)var9.getKey())) {
-               HashMap var15 = new HashMap();
-               var15.put("key", var9.getKey());
-               if (((JsonElement)var9.getValue()).isJsonPrimitive()) {
-                  var15.put("value", ((JsonElement)var9.getValue()).getAsString());
+         for (Entry<String, JsonElement> entry : messagePart.entrySet()) {
+            if (TextualComponent.isTextKey(entry.getKey())) {
+               Map<String, Object> serializedMapForm = new HashMap<>();
+               serializedMapForm.put("key", entry.getKey());
+               if (entry.getValue().isJsonPrimitive()) {
+                  serializedMapForm.put("value", entry.getValue().getAsString());
                } else {
-                  for (Entry var12 : ((JsonElement)var9.getValue()).getAsJsonObject().entrySet()) {
-                     var15.put("value." + (String)var12.getKey(), ((JsonElement)var12.getValue()).getAsString());
+                  for (Entry<String, JsonElement> compositeNestedElement : entry.getValue().getAsJsonObject().entrySet()) {
+                     serializedMapForm.put("value." + compositeNestedElement.getKey(), compositeNestedElement.getValue().getAsString());
                   }
                }
 
-               var6.text = TextualComponent.deserialize(var15);
-            } else if (MessagePart.stylesToNames.inverse().containsKey(var9.getKey())) {
-               if (((JsonElement)var9.getValue()).getAsBoolean()) {
-                  var6.styles.add((ChatColor)MessagePart.stylesToNames.inverse().get(var9.getKey()));
+               component.text = TextualComponent.deserialize(serializedMapForm);
+            } else if (MessagePart.stylesToNames.inverse().containsKey(entry.getKey())) {
+               if (entry.getValue().getAsBoolean()) {
+                  component.styles.add((ChatColor)MessagePart.stylesToNames.inverse().get(entry.getKey()));
                }
-            } else if (((String)var9.getKey()).equals("color")) {
-               var6.color = ChatColor.valueOf(((JsonElement)var9.getValue()).getAsString().toUpperCase(Locale.ROOT));
-            } else if (((String)var9.getKey()).equals("clickEvent")) {
-               JsonObject var14 = ((JsonElement)var9.getValue()).getAsJsonObject();
-               var6.clickActionName = var14.get("action").getAsString();
-               var6.clickActionData = var14.get("value").getAsString();
-            } else if (((String)var9.getKey()).equals("hoverEvent")) {
-               JsonObject var13 = ((JsonElement)var9.getValue()).getAsJsonObject();
-               var6.hoverActionName = var13.get("action").getAsString();
-               if (var13.get("value").isJsonPrimitive()) {
-                  var6.hoverActionData = new JsonString(var13.get("value").getAsString());
+            } else if (entry.getKey().equals("color")) {
+               component.color = ChatColor.valueOf(entry.getValue().getAsString().toUpperCase(Locale.ROOT));
+            } else if (entry.getKey().equals("clickEvent")) {
+               JsonObject object = entry.getValue().getAsJsonObject();
+               component.clickActionName = object.get("action").getAsString();
+               component.clickActionData = object.get("value").getAsString();
+            } else if (entry.getKey().equals("hoverEvent")) {
+               JsonObject object = entry.getValue().getAsJsonObject();
+               component.hoverActionName = object.get("action").getAsString();
+               if (object.get("value").isJsonPrimitive()) {
+                  component.hoverActionData = new JsonString(object.get("value").getAsString());
                } else {
-                  var6.hoverActionData = deserialize(var13.get("value").toString());
+                  component.hoverActionData = deserialize(object.get("value").toString());
                }
-            } else if (((String)var9.getKey()).equals("insertion")) {
-               var6.insertionData = ((JsonElement)var9.getValue()).getAsString();
-            } else if (((String)var9.getKey()).equals("with")) {
-               for (JsonElement var11 : ((JsonElement)var9.getValue()).getAsJsonArray()) {
-                  if (var11.isJsonPrimitive()) {
-                     var6.translationReplacements.add(new JsonString(var11.getAsString()));
+            } else if (entry.getKey().equals("insertion")) {
+               component.insertionData = entry.getValue().getAsString();
+            } else if (entry.getKey().equals("with")) {
+               for (JsonElement object : entry.getValue().getAsJsonArray()) {
+                  if (object.isJsonPrimitive()) {
+                     component.translationReplacements.add(new JsonString(object.getAsString()));
                   } else {
-                     var6.translationReplacements.add(deserialize(var11.toString()));
+                     component.translationReplacements.add(deserialize(object.toString()));
                   }
                }
             }
          }
 
-         var3.messageParts.add(var6);
+         returnVal.messageParts.add(component);
       }
 
-      return var3;
+      return returnVal;
    }
 
-   public FancyMessage clone() {
-      FancyMessage var1 = (FancyMessage)super.clone();
-      var1.messageParts = new ArrayList<>(this.messageParts.size());
+   public FancyMessage clone() throws CloneNotSupportedException {
+      FancyMessage instance = (FancyMessage)super.clone();
+      instance.messageParts = new ArrayList<>(this.messageParts.size());
 
-      for (int var2 = 0; var2 < this.messageParts.size(); var2++) {
-         var1.messageParts.add(var2, this.messageParts.get(var2).clone());
+      for (int i = 0; i < this.messageParts.size(); i++) {
+         instance.messageParts.add(i, this.messageParts.get(i).clone());
       }
 
-      var1.dirty = false;
-      var1.jsonString = null;
-      return var1;
+      instance.dirty = false;
+      instance.jsonString = null;
+      return instance;
    }
 
-   public FancyMessage text(String var1) {
-      MessagePart var2 = this.latest();
-      var2.text = TextualComponent.rawText(var1);
+   public FancyMessage text(String text) {
+      MessagePart latest = this.latest();
+      latest.text = TextualComponent.rawText(text);
       this.dirty = true;
       return this;
    }
 
-   public FancyMessage text(TextualComponent var1) {
-      MessagePart var2 = this.latest();
-      var2.text = var1;
+   public FancyMessage text(TextualComponent text) {
+      MessagePart latest = this.latest();
+      latest.text = text;
       this.dirty = true;
       return this;
    }
 
-   public FancyMessage color(ChatColor var1) {
-      if (!var1.isColor()) {
-         throw new IllegalArgumentException(var1.name() + " is not a color");
+   public FancyMessage color(ChatColor color) {
+      if (!color.isColor()) {
+         throw new IllegalArgumentException(color.name() + " is not a color");
       } else {
-         this.latest().color = var1;
+         this.latest().color = color;
          this.dirty = true;
          return this;
       }
    }
 
-   public FancyMessage style(ChatColor... var1) {
-      for (ChatColor var5 : var1) {
-         if (!var5.isFormat()) {
-            throw new IllegalArgumentException(var5.name() + " is not a style");
+   public FancyMessage style(ChatColor... styles) {
+      for (ChatColor style : styles) {
+         if (!style.isFormat()) {
+            throw new IllegalArgumentException(style.name() + " is not a style");
          }
       }
 
-      this.latest().styles.addAll(Arrays.asList(var1));
+      this.latest().styles.addAll(Arrays.asList(styles));
       this.dirty = true;
       return this;
    }
 
-   public FancyMessage file(String var1) {
-      this.onClick("open_file", var1);
+   public FancyMessage file(String path) {
+      this.onClick("open_file", path);
       return this;
    }
 
-   public FancyMessage link(String var1) {
-      this.onClick("open_url", var1);
+   public FancyMessage link(String url) {
+      this.onClick("open_url", url);
       return this;
    }
 
-   public FancyMessage suggest(String var1) {
-      this.onClick("suggest_command", var1);
+   public FancyMessage suggest(String command) {
+      this.onClick("suggest_command", command);
       return this;
    }
 
-   public FancyMessage insert(String var1) {
-      this.latest().insertionData = var1;
+   public FancyMessage insert(String command) {
+      this.latest().insertionData = command;
       this.dirty = true;
       return this;
    }
 
-   public FancyMessage command(String var1) {
-      this.onClick("run_command", var1);
+   public FancyMessage command(String command) {
+      this.onClick("run_command", command);
       return this;
    }
 
-   public FancyMessage achievementTooltip(String var1) {
-      this.onHover("show_achievement", new JsonString("achievement." + var1));
+   public FancyMessage achievementTooltip(String name) {
+      this.onHover("show_achievement", new JsonString("achievement." + name));
       return this;
    }
 
-   private FancyMessage tooltip(String var1) {
-      this.onHover("show_text", new JsonString(var1));
+   private FancyMessage tooltip(String text) {
+      this.onHover("show_text", new JsonString(text));
       return this;
    }
 
-   public FancyMessage tooltip(Iterable<String> var1) {
-      this.tooltip((String[])ArrayWrapper.toArray(var1, String.class));
+   public FancyMessage tooltip(Iterable<String> lines) {
+      this.tooltip((String[])ArrayWrapper.toArray(lines, String.class));
       return this;
    }
 
-   public FancyMessage tooltip(String... var1) {
-      StringBuilder var2 = new StringBuilder();
+   public FancyMessage tooltip(String... lines) {
+      StringBuilder builder = new StringBuilder();
 
-      for (int var3 = 0; var3 < var1.length; var3++) {
-         var2.append(var1[var3]);
-         if (var3 != var1.length - 1) {
-            var2.append('\n');
+      for (int i = 0; i < lines.length; i++) {
+         builder.append(lines[i]);
+         if (i != lines.length - 1) {
+            builder.append('\n');
          }
       }
 
-      this.tooltip(var2.toString());
+      this.tooltip(builder.toString());
       return this;
    }
 
-   private FancyMessage formattedTooltip(FancyMessage var1) {
-      for (MessagePart var3 : var1.messageParts) {
-         if (var3.clickActionData != null && var3.clickActionName != null) {
+   private FancyMessage formattedTooltip(FancyMessage text) {
+      for (MessagePart component : text.messageParts) {
+         if (component.clickActionData != null && component.clickActionName != null) {
             throw new IllegalArgumentException("The tooltip text cannot have click data.");
          }
 
-         if (var3.hoverActionData != null && var3.hoverActionName != null) {
+         if (component.hoverActionData != null && component.hoverActionName != null) {
             throw new IllegalArgumentException("The tooltip text cannot have a tooltip.");
          }
       }
 
-      this.onHover("show_text", var1);
+      this.onHover("show_text", text);
       return this;
    }
 
-   private FancyMessage formattedTooltip(FancyMessage... var1) {
-      if (var1.length < 1) {
+   private FancyMessage formattedTooltip(FancyMessage... lines) {
+      if (lines.length < 1) {
          this.onHover(null, null);
          return this;
       } else {
-         FancyMessage var2 = new FancyMessage();
-         var2.messageParts.clear();
+         FancyMessage result = new FancyMessage();
+         result.messageParts.clear();
 
-         for (int var3 = 0; var3 < var1.length; var3++) {
+         for (int i = 0; i < lines.length; i++) {
             try {
-               for (MessagePart var5 : var1[var3]) {
-                  if (var5.clickActionData != null && var5.clickActionName != null) {
+               for (MessagePart component : lines[i]) {
+                  if (component.clickActionData != null && component.clickActionName != null) {
                      throw new IllegalArgumentException("The tooltip text cannot have click data.");
                   }
 
-                  if (var5.hoverActionData != null && var5.hoverActionName != null) {
+                  if (component.hoverActionData != null && component.hoverActionName != null) {
                      throw new IllegalArgumentException("The tooltip text cannot have a tooltip.");
                   }
 
-                  if (var5.hasText()) {
-                     var2.messageParts.add(var5.clone());
+                  if (component.hasText()) {
+                     result.messageParts.add(component.clone());
                   }
                }
 
-               if (var3 != var1.length - 1) {
-                  var2.messageParts.add(new MessagePart(TextualComponent.rawText("\n")));
+               if (i != lines.length - 1) {
+                  result.messageParts.add(new MessagePart(TextualComponent.rawText("\n")));
                }
-            } catch (CloneNotSupportedException var6) {
+            } catch (Exception var6) {
                Bukkit.getLogger().log(Level.WARNING, "Failed to clone object", (Throwable)var6);
                return this;
             }
          }
 
-         return this.formattedTooltip(var2.messageParts.isEmpty() ? null : var2);
+         return this.formattedTooltip(result.messageParts.isEmpty() ? null : result);
       }
    }
 
-   public FancyMessage formattedTooltip(Iterable<FancyMessage> var1) {
-      return this.formattedTooltip((FancyMessage[])ArrayWrapper.toArray(var1, FancyMessage.class));
+   public FancyMessage formattedTooltip(Iterable<FancyMessage> lines) {
+      return this.formattedTooltip((FancyMessage[])ArrayWrapper.toArray(lines, FancyMessage.class));
    }
 
-   public FancyMessage translationReplacements(String... var1) {
-      for (String var5 : var1) {
-         this.latest().translationReplacements.add(new JsonString(var5));
+   public FancyMessage translationReplacements(String... replacements) {
+      for (String str : replacements) {
+         this.latest().translationReplacements.add(new JsonString(str));
       }
 
       this.dirty = true;
       return this;
    }
 
-   private FancyMessage translationReplacements(FancyMessage... var1) {
-      Collections.addAll(this.latest().translationReplacements, var1);
+   private FancyMessage translationReplacements(FancyMessage... replacements) {
+      Collections.addAll(this.latest().translationReplacements, replacements);
       this.dirty = true;
       return this;
    }
 
-   public FancyMessage translationReplacements(Iterable<FancyMessage> var1) {
-      return this.translationReplacements((FancyMessage[])ArrayWrapper.toArray(var1, FancyMessage.class));
+   public FancyMessage translationReplacements(Iterable<FancyMessage> replacements) {
+      return this.translationReplacements((FancyMessage[])ArrayWrapper.toArray(replacements, FancyMessage.class));
    }
 
-   public FancyMessage then(String var1) {
-      return this.then(TextualComponent.rawText(ColorUtils.format(var1)));
+   public FancyMessage then(String text) {
+      return this.then(TextualComponent.rawText(ColorUtils.format(text)));
    }
 
-   private FancyMessage then(TextualComponent var1) {
+   private FancyMessage then(TextualComponent text) {
       if (!this.latest().hasText()) {
          throw new IllegalStateException("previous message part has no text");
       } else {
-         this.messageParts.add(new MessagePart(var1));
+         this.messageParts.add(new MessagePart(text));
          this.dirty = true;
          return this;
       }
@@ -321,17 +321,17 @@ public class FancyMessage implements JsonRepresentedObject, Cloneable, Iterable<
       }
    }
 
-   public void writeJson(JsonWriter var1) {
+   public void writeJson(JsonWriter writer) throws IOException {
       if (this.messageParts.size() == 1) {
-         this.latest().writeJson(var1);
+         this.latest().writeJson(writer);
       } else {
-         var1.beginObject().name("text").value("").name("extra").beginArray();
+         writer.beginObject().name("text").value("").name("extra").beginArray();
 
-         for (MessagePart var3 : this) {
-            var3.writeJson(var1);
+         for (MessagePart part : this) {
+            part.writeJson(writer);
          }
 
-         var1.endArray().endObject();
+         writer.endArray().endObject();
       }
    }
 
@@ -339,84 +339,84 @@ public class FancyMessage implements JsonRepresentedObject, Cloneable, Iterable<
       if (!this.dirty && this.jsonString != null) {
          return this.jsonString;
       } else {
-         StringWriter var1 = new StringWriter();
-         JsonWriter var2 = new JsonWriter(var1);
+         StringWriter string = new StringWriter();
+         JsonWriter json = new JsonWriter(string);
 
          try {
-            this.writeJson(var2);
-            var2.close();
+            this.writeJson(json);
+            json.close();
          } catch (IOException var4) {
             throw new RuntimeException("invalid message");
          }
 
-         this.jsonString = var1.toString();
+         this.jsonString = string.toString();
          this.dirty = false;
          return this.jsonString;
       }
    }
 
-   public void send(Player var1) {
-      this.send(var1, this.toJSONString());
+   public void send(Player player) {
+      this.send(player, this.toJSONString());
    }
 
-   private void send(CommandSender var1, String var2) {
-      if (!(var1 instanceof Player var3)) {
-         var1.sendMessage(this.toOldMessageFormat());
+   private void send(CommandSender sender, String jsonString) {
+      if (!(sender instanceof Player player)) {
+         sender.sendMessage(this.toOldMessageFormat());
       } else {
-         Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + var3.getName() + " " + var2);
+         player.sendMessage(net.kyori.adventure.text.serializer.gson.GsonComponentSerializer.gson().deserialize(jsonString));
       }
    }
 
-   public void send(CommandSender var1) {
-      this.send(var1, this.toJSONString());
+   public void send(CommandSender sender) {
+      this.send(sender, this.toJSONString());
    }
 
-   public void send(Iterable<? extends CommandSender> var1) {
-      String var2 = this.toJSONString();
+   public void send(Iterable<? extends CommandSender> senders) {
+      String string = this.toJSONString();
 
-      for (CommandSender var4 : var1) {
-         this.send(var4, var2);
+      for (CommandSender sender : senders) {
+         this.send(sender, string);
       }
    }
 
    private String toOldMessageFormat() {
-      StringBuilder var1 = new StringBuilder();
+      StringBuilder result = new StringBuilder();
 
-      for (MessagePart var3 : this) {
-         var1.append(var3.color == null ? "" : var3.color);
+      for (MessagePart part : this) {
+         result.append(part.color == null ? "" : part.color);
 
-         for (ChatColor var5 : var3.styles) {
-            var1.append(var5);
+         for (ChatColor formatSpecifier : part.styles) {
+            result.append(formatSpecifier);
          }
 
-         var1.append(var3.text);
+         result.append(part.text);
       }
 
-      return var1.toString();
+      return result.toString();
    }
 
    private MessagePart latest() {
       return this.messageParts.get(this.messageParts.size() - 1);
    }
 
-   private void onClick(String var1, String var2) {
-      MessagePart var3 = this.latest();
-      var3.clickActionName = var1;
-      var3.clickActionData = var2;
+   private void onClick(String name, String data) {
+      MessagePart latest = this.latest();
+      latest.clickActionName = name;
+      latest.clickActionData = data;
       this.dirty = true;
    }
 
-   private void onHover(String var1, JsonRepresentedObject var2) {
-      MessagePart var3 = this.latest();
-      var3.hoverActionName = var1;
-      var3.hoverActionData = var2;
+   private void onHover(String name, JsonRepresentedObject data) {
+      MessagePart latest = this.latest();
+      latest.hoverActionName = name;
+      latest.hoverActionData = data;
       this.dirty = true;
    }
 
    public Map<String, Object> serialize() {
-      HashMap var1 = new HashMap();
-      var1.put("messageParts", this.messageParts);
-      return var1;
+      HashMap<String, Object> map = new HashMap<>();
+      map.put("messageParts", this.messageParts);
+      return map;
    }
 
    @Override

